[onMobDeath] Replace isKiller and noKiller with optParams table

claywar opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?
* Replaces isKiller and noKiller onMobDeath parameters with optParams table.  This will allow for additional parameters to be added with no impact to existing lua scripts.
* Adds isWeaponSkillKill to optParams table.

Returned format is currently (all boolean values):
```
optParams =
{
    isKiller,
    noKiller,
    isWeaponSkillKill
}
```
<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

## Steps to test these changes
Test procedure used:
1. Modify mob script onMobDeath with the following:
```
print(optParams.isKiller)
print(optParams.noKiller)
print(optParams.isWeaponSkillKill)
```
2. Go to modified mob, kill to get the above conditions to change.

<!-- Clear and detailed steps to test your changes here -->

claywar committed:
[core] Add optParams table to onMobDeath, remove isKiller, noKiller

claywar committed:
[lua] Replace (mob, player, isKiller) with (mob, player, optParams)

claywar committed:
[lua] Replace (mob, player, isKiller, noKiller) with (mob, player, optParams)

claywar committed:
[lua] Change references from noKiller to optParams.noKiller

zach2good commented:
Diff is too large for me to easily get to what I want, but:
```diff
diff --git a/src/map/lua/luautils.cpp b/src/map/lua/luautils.cpp
index 631e0b7cd52..4835f4ee962 100644
--- a/src/map/lua/luautils.cpp
+++ b/src/map/lua/luautils.cpp
@@ -3223,7 +3223,8 @@ namespace luautils
         std::string zone_name = (const char*)PMob->loc.zone->GetName();
         std::string mob_name  = (const char*)PMob->GetName();
 
-        CCharEntity* PChar = dynamic_cast<CCharEntity*>(PKiller);
+        CCharEntity* PChar     = dynamic_cast<CCharEntity*>(PKiller);
+        auto         optParams = lua.create_table();
         if (PChar && PMob->objtype == TYPE_MOB)
         {
             auto onMobDeathEx = lua["xi"]["mob"]["onMobDeathEx"];
@@ -3258,7 +3259,7 @@ namespace luautils
             sol::function onMobDeath          = getEntityCachedFunction(PMob, "onMobDeath");
 
             // clang-format off
-            PChar->ForAlliance([PMob, PChar, &onMobDeathFramework, &onMobDeath, &filename](CBattleEntity* PPartyMember)
+            PChar->ForAlliance([PMob, PChar, &onMobDeathFramework, &onMobDeath, &filename, &optParams](CBattleEntity* PPartyMember)
             {
                 CCharEntity* PMember = (CCharEntity*)PPartyMember;
                 if (PMember && PMember->getZone() == PChar->getZone())
@@ -3269,14 +3270,16 @@ namespace luautils
                     {
                         optLuaAllyEntity = CLuaBaseEntity(PMember);
                     }
-                    bool isKiller = PMember == PChar;
-                    bool noKiller = false;
+
+                    optParams["isKiller"]          = PMember == PChar;
+                    optParams["noKiller"]          = false;
+                    optParams["isWeaponSkillKill"] = PMob->GetLocalVar("weaponskillHit") > 0;
 
                     PChar->eventPreparation->targetEntity = PMob;
                     PChar->eventPreparation->scriptFile   = filename;
 
                     // onMobDeath(mob, player, isKiller, noKiller)
-                    auto result = onMobDeathFramework(LuaMobEntity, optLuaAllyEntity, isKiller, noKiller, onMobDeath);
+                    auto result = onMobDeathFramework(LuaMobEntity, optLuaAllyEntity, optParams, onMobDeath);
                     if (!result.valid())
                     {
                         sol::error err = result;
@@ -3291,8 +3294,12 @@ namespace luautils
             auto          onMobDeathFramework = lua["xi"]["globals"]["interaction"]["interaction_global"]["onMobDeath"];
             sol::function onMobDeath          = getEntityCachedFunction(PMob, "onMobDeath");
 
+            optParams["isKiller"]          = false;
+            optParams["noKiller"]          = true;
+            optParams["isWeaponSkillKill"] = false;
+
             // onMobDeath(mob, player, isKiller, noKiller)
-            auto result = onMobDeathFramework(CLuaBaseEntity(PMob), sol::lua_nil, sol::lua_nil, true, onMobDeath);
+            auto result = onMobDeathFramework(CLuaBaseEntity(PMob), sol::lua_nil, optParams, onMobDeath);
             if (!result.valid())
             {
                 sol::error err = result;
```

Could this section (default values) be moved up to just after the creation of the optParams table?
```
 
+            optParams["isKiller"]          = false;
+            optParams["noKiller"]          = true;
+            optParams["isWeaponSkillKill"] = false;
```

claywar committed:
[core] Create optParams only once, change isKiller to false in noKiller scenario from sol::lua_nil


CI - multiple newlines in todo files


Move default values out of condition


Formatting
